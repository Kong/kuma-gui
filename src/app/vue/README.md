---
section: Packages
---
# vue

Automatic `vue` initialization using our projects dependency injection layer.

