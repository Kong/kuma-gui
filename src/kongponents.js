import Vue from 'vue'
import '@kongponents/styles/styles.css'
import KAlert from '@kongponents/kalert'
import KBadge from '@kongponents/kbadge'
import KButton from '@kongponents/kbutton'
import KCard from '@kongponents/kcard'
import KEmptyState from '@kongponents/kemptystate'
import KIcon from '@kongponents/kicon'
import KTable from '@kongponents/ktable'
import KToaster from '@kongponents/ktoaster'
import KoolTip from '@kongponents/kooltip'
import KClipboardProvider from '@kongponents/kclipboardprovider'
import Krumbs from '@kongponents/krumbs'
import KPop from '@kongponents/kpop'
import KToggle from '@kongponents/ktoggle'
import KTabs from '@kongponents/ktabs'
import KModal from '@kongponents/kmodal'
import KInput from '@kongponents/kinput'
import KInputSwitch from '@kongponents/kinputswitch'

Vue.component('KAlert', KAlert)
Vue.component('KBadge', KBadge)
Vue.component('KButton', KButton)
Vue.component('KCard', KCard)
Vue.component('KEmptyState', KEmptyState)
Vue.component('KIcon', KIcon)
Vue.component('KTable', KTable)
Vue.component('KToaster', KToaster)
Vue.component('KoolTip', KoolTip)
Vue.component('KClipboardProvider', KClipboardProvider)
Vue.component('Krumbs', Krumbs)
Vue.component('KPop', KPop)
Vue.component('KTabs', KTabs)
Vue.component('KToggle', KToggle)
Vue.component('KInput', KInput)
Vue.component('KInputSwitch', KInputSwitch)
Vue.component('KModal', KModal)
