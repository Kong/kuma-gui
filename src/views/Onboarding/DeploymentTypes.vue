<template>
  <OnboardingPage>
    <template #content>
      <OnboardingHeading
        title="Deployment Types"
        :description="
          `To get started with ${title}, we must first take a look at a few basic concepts.`
        "
      />

      <HoverableSvgWrapper>
        <component :is="currentGraph" />
      </HoverableSvgWrapper>
      <KRadio
        v-model="mode"
        name="mode"
        value="standalone"
      >
        Standalone Deployment
      </KRadio>
      <KRadio
        v-model="mode"
        name="mode"
        value="multi-zone"
      >
        Multi-Zone deployment
      </KRadio>
    </template>

    <template #navigation>
      <OnboardingNavigation
        next-step="onboarding-backend-types"
        previous-step="onboarding-welcome"
      />
    </template>
  </OnboardingPage>
</template>

<script>
import { mapGetters } from 'vuex'
import MultiZoneDeploymentGraph from '@/views/Onboarding/components/MultiZoneDeploymentGraph'
import StandaloneDeploymentGraph from '@/views/Onboarding/components/StandaloneDeploymentGraph'
import OnboardingNavigation from '@/views/Onboarding/components/OnboardingNavigation'
import OnboardingHeading from '@/views/Onboarding/components/OnboardingHeading'
import HoverableSvgWrapper from '@/views/Onboarding/components/HoverableSvgWrapper'
import OnboardingPage from '@/views/Onboarding/components/OnboardingPage'

export default {
  name: 'DeploymentTypes',
  components: {
    MultiZoneDeploymentGraph,
    StandaloneDeploymentGraph,
    OnboardingNavigation,
    OnboardingHeading,
    HoverableSvgWrapper,
    OnboardingPage,
  },
  data() {
    return { mode: 'standalone' }
  },

  computed: {
    ...mapGetters({
      title: 'config/getTagline',
      multicluster: 'config/getMulticlusterStatus',
    }),
    currentGraph() {
      return this.mode === 'standalone' ? 'StandaloneDeploymentGraph' : 'MultiZoneDeploymentGraph'
    },
  },
  mounted() {
    this.mode = this.multicluster ? 'multi-zone' : 'standalone'
  },
}
</script>
