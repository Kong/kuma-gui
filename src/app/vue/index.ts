import { createApp } from 'vue'
import {
  createRouter,
  createWebHistory,
} from 'vue-router'

import type { ServiceDefinition } from '@/services/utils'
import { token, createInjections } from '@/services/utils'
import type { Component } from 'vue'
import type { Router, RouteRecordRaw, NavigationGuard } from 'vue-router'

type Token = ReturnType<typeof token>

type VueApp = ReturnType<typeof createApp>
export type PluginDefinition = Parameters<VueApp['use']>
export type ComponentDefinition = [string, Component]

const $ = {
  app: token<(App: Component) => Promise<VueApp>>('vue.app'),
  router: token<Router>('vue.router'),

  components: token('vue.components'),
  plugins: token<PluginDefinition[]>('vue.plugins'),
  routes: token<RouteRecordRaw[]>('vue.routes'),
  routesLabel: token<RouteRecordRaw[]>('vue.routes.label'),
  navigationGuards: token<NavigationGuard[]>('vue.routes.navigation.guards'),
}
export const services = (app: Record<string, Token>): ServiceDefinition[] => {
  return [
    [$.app, {
      service: (
        components: ComponentDefinition[],
        plugins: PluginDefinition[],
      ) => {
        return async (App: Component) => {
          const app = createApp(App)

          plugins.forEach(([...args]) => {
            app.use(...args)
          })

          components.forEach(([name, component]: [string, Component]) => {
            app.component(name, component)
          })

          return app
        }
      },
      arguments: [
        $.components,
        $.plugins,
      ],
    }],
    [$.router, {
      service: (
        env: (str: 'KUMA_BASE_PATH') => string,
        routes: RouteRecordRaw[],
        guards: NavigationGuard[],
      ) => {
        const router = createRouter({
          history: createWebHistory(env('KUMA_BASE_PATH')),
          routes,
        })

        guards.forEach((item) => {
          if (typeof item === 'function') {
            router.beforeEach(item)
          }
        })

        return router
      },
      arguments: [
        app.env,
        $.routes,
        $.navigationGuards,
      ],
    }],

    [token('vue.plugins'), {
      service: (router) => {
        return [
          [router],
        ]
      },
      arguments: [
        $.router,
      ],
      labels: [
        $.plugins,
      ],
    }],

    // TODO(jc): Label decoration. Make it so we can decorate labels
    // Temporarily turn the routesLabel label into routes for decoration
    // purposes
    [$.routes, {
      service: (routes: RouteRecordRaw[]) => routes,
      arguments: [
        $.routesLabel,
      ],
    }],
    //

    // TODO(jc): Make it so we don't need to provide a [] empty default for
    // labels
    [token('application.routes.navigation.guards'), {
      service: () => {
        return []
      },
      labels: [
        $.navigationGuards,
      ],
    }],
  ]
}

export const TOKENS = $
export const [
  useRouter,
] = createInjections(
  TOKENS.router,
)
