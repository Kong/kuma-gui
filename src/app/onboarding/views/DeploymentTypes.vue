<template>
  <RouteView>
    <RouteTitle
      :title="t('onboarding.routes.deployment-types.title')"
    />
    <AppView>
      <OnboardingPage with-image>
        <template #header>
          <OnboardingHeading>
            <template #title>
              Learn about deployments
            </template>

            <template #description>
              <p>{{ t('common.product.name') }} can be deployed in standalone or multi-zone mode.</p>
            </template>
          </onboardingheading>
        </template>

        <template #content>
          <div class="graph-list mb-6">
            <component :is="currentGraphComponent" />
          </div>

          <div class="radio-button-group">
            <KRadio
              v-model="mode"
              name="mode"
              selected-value="standalone"
              data-testid="onboarding-standalone-radio-button"
            >
              Standalone deployment
            </KRadio>

            <KRadio
              v-model="mode"
              name="mode"
              selected-value="multi-zone"
              data-testid="onboarding-multi-zone-radio-button"
            >
              Multi-zone deployment
            </KRadio>
          </div>
        </template>

        <template #navigation>
          <OnboardingNavigation
            next-step="onboarding-configuration-types"
            previous-step="onboarding-welcome"
          />
        </template>
      </OnboardingPage>
    </AppView>
  </RouteView>
</template>

<script lang="ts" setup>
import { KRadio } from '@kong/kongponents'
import { computed, onMounted, ref } from 'vue'

import OnboardingHeading from '../components/OnboardingHeading.vue'
import OnboardingNavigation from '../components/OnboardingNavigation.vue'
import OnboardingPage from '../components/OnboardingPage.vue'
import AppView from '@/app/application/components/app-view/AppView.vue'
import RouteTitle from '@/app/application/components/route-view/RouteTitle.vue'
import RouteView from '@/app/application/components/route-view/RouteView.vue'
import {
  useMultizoneGraph,
  useStandaloneGraph,
} from '@/components'
import { useStore } from '@/store/store'
import { useI18n } from '@/utilities'

const MultizoneGraph = useMultizoneGraph()
const StandaloneGraph = useStandaloneGraph()

const componentMap: Record<string, any> = {
  standalone: StandaloneGraph,
  'multi-zone': MultizoneGraph,
}

const store = useStore()
const { t } = useI18n()

const mode = ref<'standalone' | 'multi-zone'>('standalone')

const currentGraphComponent = computed(() => componentMap[mode.value])

onMounted(function () {
  mode.value = store.getters['config/getMulticlusterStatus'] ? 'multi-zone' : 'standalone'
})
</script>

<style lang="scss" scoped>
.graph-list {
  width: 100%;
  height: 100%;
  display: flex;
  justify-content: space-evenly;
  align-items: center;
}

.radio-button-group {
  display: flex;
  justify-content: center;
  gap: $kui-space-80;
  margin-bottom: $kui-space-60;
}

.radio-button-group .k-radio {
  cursor: pointer;
}
</style>
