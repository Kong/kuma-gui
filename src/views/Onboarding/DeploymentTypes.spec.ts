import { createStore } from 'vuex'
import { RouterLinkStub } from '@vue/test-utils'
import userEvent from '@testing-library/user-event'
import { render, screen } from '@testing-library/vue'
import { KButton, KPop, KRadio } from '@kong/kongponents'

import DeploymentTypes from './DeploymentTypes.vue'

function renderComponent() {
  const store = createStore({
    modules: {
      config: {
        namespaced: true,
        state: {
          clientConfig: {
            mode: 'global',
            store: {
              type: 'memory',
            },
          },
        },
        getters: {
          getConfigurationType: (state) => state.clientConfig?.store?.type,
          getMulticlusterStatus: () => true,
        },
      },
    },
  })

  return render(DeploymentTypes, {
    global: {
      plugins: [store],
      components: {
        KButton,
        KPop,
        KRadio,
      },
      stubs: {
        routerLink: RouterLinkStub,
      },
    },
  })
}

describe('DeploymentTypes.vue', () => {
  it('renders snapshot', () => {
    const { container } = renderComponent()

    expect(container).toMatchSnapshot()
  })

  it('changes selected graph', async () => {
    renderComponent()

    await userEvent.click(screen.getByText(/Multi-zone deployment/))

    expect(screen.getByTestId('multizone-graph')).toBeInTheDocument()
  })
})
