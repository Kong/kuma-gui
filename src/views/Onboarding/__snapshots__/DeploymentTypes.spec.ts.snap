// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DeploymentTypes.vue renders snapshot 1`] = `
<div>
  <div
    class="onboarding-container"
  >
    <div
      class="min-h-96"
    >
      <div
        class="mb-6"
      >
        <h1
          class="text-center text-4xl mb-4 min-h-8"
        >
          
    Deployment Types
  
        </h1>
         
        <p
          class="text-center"
        >
          
    To get started with Kuma, we must first take a look at a few basic concepts.
  
        </p>
      </div>
       
      <div
        class="w-full relative "
      >
        <transition-stub
          leaveactiveclass="transition-opacity duration-300"
          leavetoclass="opacity-0"
        >
          <div
            class="absolute bg-black bg-opacity-75 w-full h-full"
            data-testid="hoverable-overlay"
          >
            <span
              class="absolute right-0 text-white mr-4 mt-4"
            >
              
        Mouse over the boxes
      
            </span>
          </div>
        </transition-stub>
         
        <div>
          <div>
            <div
              data-v-3cb990b6=""
            >
              <span
                data-v-3cb990b6=""
              >
                Standalone 1
              </span>
              <transition-stub
                data-v-3cb990b6=""
                name="fade"
              >
                <div
                  class="k-popover"
                  data-v-3cb990b6=""
                  style="width: 200px; display: none;"
                >
                  <!---->
                  <div
                    class="k-popover-content"
                    data-v-3cb990b6=""
                  >
                    <p>
                      Text of Standalone 1
                    </p>
                  </div>
                </div>
              </transition-stub>
            </div>
          </div>
           
          <svg
            height="151px"
            version="1.1"
            viewBox="-0.5 -0.5 441 151"
            width="441px"
            xmlns="http://www.w3.org/2000/svg"
            xmlns:xlink="http://www.w3.org/1999/xlink"
          >
            <defs />
             
            <g>
              <g
                data-v-3cb990b6=""
              >
                <g
                  data-v-3cb990b6=""
                >
                  <rect
                    data-v-3cb990b6=""
                    fill="#00cc00"
                    height="60"
                    pointer-events="all"
                    stroke="#000000"
                    width="120"
                    x="160"
                    y="0"
                  />
                   
                  <text
                    data-v-3cb990b6=""
                    fill="#000000"
                    font-family="Helvetica"
                    font-size="12px"
                    text-anchor="middle"
                    x="220"
                    y="34"
                  >
                    
            Kuma control plane
          
                  </text>
                </g>
                <div
                  data-v-3cb990b6=""
                  name="fade"
                >
                  <foreignobject
                    data-v-3cb990b6=""
                  >
                    <div
                      class="k-popover"
                      data-v-3cb990b6=""
                      style="width: 200px; display: none;"
                    >
                      <div
                        class="k-popover-title"
                        data-v-3cb990b6=""
                      >
                        Light
                      </div>
                      <div
                        class="k-popover-content"
                        data-v-3cb990b6=""
                      >
                        <p
                          data-v-3cb990b6=""
                        >
                          Text of Control Plane
                        </p>
                      </div>
                    </div>
                  </foreignobject>
                </div>
              </g>
               
              <rect
                fill="#ffffff"
                height="60"
                pointer-events="all"
                stroke="#000000"
                width="120"
                x="0"
                y="90"
              />
               
              <text
                fill="#000000"
                font-family="Helvetica"
                font-size="12px"
                text-anchor="middle"
                x="40"
                y="124"
              >
                
        Service
      
              </text>
               
              <rect
                fill="#ffffff"
                height="60"
                pointer-events="all"
                stroke="#000000"
                width="120"
                x="320"
                y="90"
              />
               
              <text
                fill="#000000"
                font-family="Helvetica"
                font-size="12px"
                text-anchor="middle"
                x="400"
                y="124"
              >
                
        Service
      
              </text>
               
              <g>
                <path
                  d="M 125.09 86.18 L 154.91 63.82"
                  fill="none"
                  pointer-events="stroke"
                  stroke="#000000"
                  stroke-miterlimit="10"
                />
                 
                <path
                  d="M 120.89 89.33 L 124.39 82.33 L 125.09 86.18 L 128.59 87.93 Z"
                  fill="#000000"
                  pointer-events="all"
                  stroke="#000000"
                  stroke-miterlimit="10"
                />
                 
                <path
                  d="M 159.11 60.67 L 155.61 67.67 L 154.91 63.82 L 151.41 62.07 Z"
                  fill="#000000"
                  pointer-events="all"
                  stroke="#000000"
                  stroke-miterlimit="10"
                />
              </g>
               
              <g>
                <path
                  d="M 314.91 86.18 L 285.09 63.82"
                  fill="none"
                  pointer-events="stroke"
                  stroke="#000000"
                  stroke-miterlimit="10"
                />
                 
                <path
                  d="M 319.11 89.33 L 311.41 87.93 L 314.91 86.18 L 315.61 82.33 Z"
                  fill="#000000"
                  pointer-events="all"
                  stroke="#000000"
                  stroke-miterlimit="10"
                />
                 
                <path
                  d="M 280.89 60.67 L 288.59 62.07 L 285.09 63.82 L 284.39 67.67 Z"
                  fill="#000000"
                  pointer-events="all"
                  stroke="#000000"
                  stroke-miterlimit="10"
                />
              </g>
               
              <g>
                <rect
                  fill="#ff3333"
                  height="60"
                  pointer-events="all"
                  stroke="#000000"
                  width="40"
                  x="80"
                  y="90"
                />
                 
                <text
                  fill="#000000"
                  font-family="Helvetica"
                  font-size="12px"
                  text-anchor="middle"
                  x="100"
                  y="124"
                >
                  
          DPP
        
                </text>
              </g>
               
              <g>
                <rect
                  fill="#ff0000"
                  height="60"
                  pointer-events="all"
                  stroke="#000000"
                  width="40"
                  x="320"
                  y="90"
                />
                 
                <text
                  fill="#000000"
                  font-family="Helvetica"
                  font-size="12px"
                  text-anchor="middle"
                  x="340"
                  y="124"
                >
                  
          DPP
        
                </text>
              </g>
            </g>
          </svg>
        </div>
      </div>
       
      <label
        class="k-radio d-inline-block"
        data-v-b7148f78=""
        name="mode"
      >
        <input
          class="k-input float-left"
          data-v-b7148f78=""
          name="mode"
          type="radio"
        />
        <span
          class="float-left"
          data-v-b7148f78=""
        >
          
      Standalone Deployment
    
        </span>
      </label>
       
      <label
        class="k-radio d-inline-block"
        data-v-b7148f78=""
        name="mode"
      >
        <input
          class="k-input float-left"
          data-v-b7148f78=""
          name="mode"
          type="radio"
        />
        <span
          class="float-left"
          data-v-b7148f78=""
        >
          
      Multi-Zone deployment
    
        </span>
      </label>
    </div>
     
    <div
      class="mt-4 flex justify-between"
    >
      <div>
        <a
          aria-current="page"
          class="k-button mr-4 router-link-exact-active router-link-active secondary"
          data-v-38fa8ecf=""
          href="#/"
          type="button"
        >
          
      Back
    
        </a>
         
        <a
          aria-current="page"
          class="k-button skip-button router-link-exact-active router-link-active small btn-link"
          data-v-38fa8ecf=""
          href="#/"
          type="button"
        >
          
      or Skip
    
        </a>
      </div>
       
      <a
        aria-current="page"
        class="k-button router-link-exact-active router-link-active primary"
        data-v-38fa8ecf=""
        href="#/"
        type="button"
      >
        
    Next
  
      </a>
    </div>
  </div>
</div>
`;
